src/lib/voiceSearchHelper/index.ts

```diff
@@ -0,0 +1,123 @@
+const STATUS_INITIAL = 'initial';
+const STATUS_ASKING_PERMISSION = 'askingPermission';
+const STATUS_WAITING = 'waiting';
+const STATUS_RECOGNIZING = 'recognizing';
+const STATUS_FINISHED = 'finished';
+const STATUS_ERROR = 'error';
+
+export type VoiceSearchHelperParams = {
+  searchAsYouSpeak: boolean;
+  onQueryChange: (query: string) => void;
+  onStateChange: () => void;
+};
+
+export type VoiceListeningState = {
+  status: string;
+  transcript?: string;
+  isSpeechFinal?: boolean;
+  errorCode?: string;
+};
+
+export type ToggleListening = () => void;
+
+export default function voiceSearchHelper({
+  searchAsYouSpeak,
+  onQueryChange,
+  onStateChange,
+}: VoiceSearchHelperParams) {
+  const SpeechRecognitionAPI: new () => SpeechRecognition =
+    (window as any).webkitSpeechRecognition ||
+    (window as any).SpeechRecognition;
+  const getDefaultState = (status: string): VoiceListeningState => ({
+    status,
+    transcript: undefined,
+    isSpeechFinal: undefined,
+    errorCode: undefined,
+  });
+  let state: VoiceListeningState = getDefaultState(STATUS_INITIAL);
+  let recognition: SpeechRecognition | undefined;
+
+  const isBrowserSupported = () => Boolean(SpeechRecognitionAPI);
+
+  const isListening = () =>
+    state.status === STATUS_ASKING_PERMISSION ||
+    state.status === STATUS_WAITING ||
+    state.status === STATUS_RECOGNIZING;
+
+  const setState = (newState = {}) => {
+    state = { ...state, ...newState };
+    onStateChange();
+  };
+
+  const getState = (): VoiceListeningState => state;
+
+  const resetState = (status = STATUS_INITIAL) => {
+    setState(getDefaultState(status));
+  };
+
+  const stop = () => {
+    if (recognition) {
+      recognition.stop();
+      recognition = undefined;
+    }
+    resetState();
+  };
+
+  const start = () => {
+    recognition = new SpeechRecognitionAPI();
```

Comment:
Are there any drawbacks at recreating a new speech recognition every time time we start listening?

Reply:
I have no idea. I haven't read on the [spec](https://w3c.github.io/speech-api) anything about reusing the instance, cleaning it up, etc. I guess recreating a new speech recognition shouldn't be a burden.